j2v8-debugger/src/main/java/com/alexii/j2v8debugger/V8Messenger.kt

```diff
@@ -0,0 +1,193 @@
+            chromeMessageQueue[responseMethod] = updatedScript
+        }
+    }
+
+    private fun handleScriptParsedEvent(responseParams: JSONObject?) {
+        val scriptParsedEvent = dtoMapper.convertValue(responseParams, ScriptParsedEventRequest::class.java)
+        if (scriptParsedEvent.url.isNotEmpty()) {
+            // Get the V8 Script ID to map to the Chrome ScriptId (stored in url)
+            v8ScriptMap[scriptParsedEvent.scriptId] = scriptParsedEvent.url
+        }
+    }
+
+    /**
+     * For BreakpointResolved events, we need to convert the scriptId from the J2V8 scriptId
+     * to the Chrome DevTools scriptId before passing it through
+     */
+    private fun handleBreakpointResolvedEvent(responseParams: JSONObject?, responseMethod: String?) {
+        val breakpointResolvedEvent = dtoMapper.convertValue(responseParams, BreakpointResolvedEvent::class.java)
+        val location = breakpointResolvedEvent.location
+        val response = BreakpointResolvedEvent().also {
+            it.breakpointId = breakpointResolvedEvent.breakpointId
+            it.location = LocationResponse().also {
+                it.scriptId = v8ScriptMap[location?.scriptId]
+                it.lineNumber = location?.lineNumber
+                it.columnNumber = location?.columnNumber
+            }
+        }
+        chromeMessageQueue[responseMethod] = dtoMapper.convertValue(response, JSONObject::class.java)
+    }
+
+    /**
+     * Send messages to J2V8
+     * If debugger is paused, they will be queued to send in [waitFrontendMessageOnPause]
+     * otherwise we can send now.
+     * Some messages are only relevant while paused so ignore them if it's not
+     */
+    fun sendMessage(message: String, params: JSONObject? = null, runOnlyWhenPaused: Boolean = false) {
+        if (debuggerState == DebuggerState.Paused) {
+            v8MessageQueue[message] = params
+        } else if (!runOnlyWhenPaused) {
+            dispatchMessage(message, params.toString())
+        }
+    }
+
+    /**
+     * Change debugger state when DevTools connects and disconnects
+     */
+    fun setDebuggerConnected(isConnected: Boolean) {
+        debuggerState = if (isConnected) DebuggerState.Connected else DebuggerState.Disconnected
+    }
+
+    /**
+     * Pass message to J2V8
+     * If we're awaiting a response in the pendingMessageQueue, use the Id and set to pending
+     */
+    private fun dispatchMessage(method: String, params: String? = null) {
+        val messageId: Int
+        val pendingMessage = pendingMessageQueue.firstOrNull { msg -> msg.method == method && !msg.pending }
+        if (pendingMessage != null) {
+            pendingMessage.pending = true
+            messageId = pendingMessage.messageId
+        } else {
+            messageId = nextDispatchId.incrementAndGet()
+        }
+        val message = "{\"id\":$messageId,\"method\":\"$method\", \"params\": ${params ?: "{}"}}"
```

Comment:
Just wondered why do we create json string manually here instead of creating json object and having it stringified version? 
Generally such string concatenations are more error prone.

Reply:
This was to prevent the overhead from going from string -> JSONObject -> string, since input is a string, and output is a string.  In general there's a lot of converting back and forth from JSONObject and string, although most of the time it's for tiny payloads.  I'm ok with switching this to a JSONObject for readability and reduce chances for bugs.

Reply:
thanks for the info and implementing the change. it's fair point regarding reducing conversion. 
Also unless we figure-out specific performance issue I would prefer readability and reduce chances for bugs.